### ISSUE TITLE
training data were keeping changing and clone did not protect it

### ISSUE THREAD

**Issue by santisy** [torch/nn#786]
As the title says, when training a cnn, the input data and target data were keeping changing, and clone did not protect it.

The code is here: [problem repository](https://github.com/santisy/problem)

In the code above, I cloned all data before starting training but not working.

Please run `main.lua` to see the two `print` are printing different values.

`netWorks` is for building cnn, `netOptim` contains the training process and `customLayers` is wrapped function for adding custom weights and bias.

**Comment by albanD** [comment]
Hi,

I think you do not use the plain `BatchNormalization` module since your code sample does not work as is. To make it work, I have to change the parameters when creating the batch norm layers to match the number of inputs. After this change, you code run as expected and the tensor is not modified afterwards.
You should avoid using globals in your project because this can easily lead to the kind of problem you see here.

**Comment by santisy** [comment]
@albanD thanks for replying. 
I indeed have had a lot of trouble with the function `BatchNormalization`. I have noticed that when run those code in CPU,`BatchNormaliztion` report errors but in GPU it doesn't. And I do not get your method, how should I change the parameters of `BatchNormalization`?  
Still I think this may not be a problem of global variables, even if you change it all to local this can happen. And I think the `clone` here dose not work is still strange, is not?

**Comment by santisy** [comment]
@albanD Thanks so much!
I have got your idea.

**Comment by fmassa** [comment]
@santisy this PR added the cuda check https://github.com/torch/nn/pull/777

**Comment by albanD** [comment]
@santisy the clone works as expected when I run your code sample: the tensor is the same before and after.
